explorer/interpreter/stack_space.h

```diff
@@ -0,0 +1,61 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_EXPLORER_INTERPRETER_STACK_SPACE_H_
+#define CARBON_EXPLORER_INTERPRETER_STACK_SPACE_H_
+
+#include <optional>
+
+#include "llvm/ADT/STLFunctionalExtras.h"
+
+namespace Carbon {
+
+namespace Internal {
+
+// Returns true if a new thread should be started for more stack space.
+auto IsStackSpaceNearlyExhausted() -> bool;
+
+// Starts a thread to run the function.
+auto RunWithStackSpaceHelper(llvm::function_ref<void()> fn) -> void;
+
+}  // namespace Internal
+
+// Initializes stack space information for RunWithStackSpace by creating a new
+// thread, and runs the function on the new thread. This only needs to be called
+// once within a given recursion.
+//
+// Usage:
+//   return InitStackSpace<ReturnType>([&]() -> ReturnType {
+//         <function body>
+//       });
+template <typename ReturnType>
+auto InitStackSpace(llvm::function_ref<ReturnType()> fn) -> ReturnType {
+  std::optional<ReturnType> result;
+  Internal::RunWithStackSpaceHelper([&] { result = fn(); });
+  return std::move(*result);
+}
+
+// Runs a function. May start a new thread for the function if too much stack
+// space has been consumed since the last time a thread was spawned (by either
+// InitStackSpace or RunWithStackSpace). Require InitStackSpace be used first
+// within the recursion.
```

Comment:
Could we drop this requirement by having this function behave like `InitStackSpace` if `bottom_of_stack == 0`? That way we wouldn't need two separate functions.

Reply:
The current approach is helpful to prevent programming errors, which would otherwise be subtle and difficult to catch.

For example, if we weren't doing this validation, what you could end up with is someone calling type checking without hitting RunWithStackSpace first, and as a consequence getting one thread per TypeCheckExp call. That would technically work, but it would also be slower than intended and it's hard to catch when excess threads are being created (I had a bug creating too many threads while implementing, so I'm perhaps extra cautious on this).

This approach addresses that by making it an error with a clear fix, calling InitStackSpace first. It pushes callers towards supported entry points.

Reply:
Those bugs seem like they will be very rare -- they would only come up when you're calling this function from an an inner loop that's not nested inside the `ParseFileAndExecute` call that wraps nearly the entire program. If so, and given the fact that they're performance bugs rather than correctness bugs, I'm not convinced that preventing them is worth complicating this API by having two subtly-different functions that callers need to choose between.

Reply:
So, the best I can say here is that I feel defensive coding is appropriate here based on my own experiences of hitting these bugs. Note, performance can be severely impacted if too many threads are started, to the extent that it becomes a correctness issue, just not one that crashes in a typical manner.

I don't know how to convince you that this is worth a separate function. Maybe it'd be good to get a separate opinion here?

Reply:
Yes, I think a second opinion is probably the best way to unblock this.

Reply:
Noting that other potential voices are now on vacation, I've made your requested change. I may follow up separately. Can you please take another look?